File: targettests/Remote-Sim/sample.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// REQUIRES: c++20
// REQUIRES: remote-sim
// FIXME: https://github.com/NVIDIA/cuda-quantum/issues/1111

// clang-format off
// RUN: nvq++ --target remote-mqpu --remote-mqpu-auto-launch 1 %s -o %t && %t 
// RUN: nvq++ --enable-mlir --target remote-mqpu --remote-mqpu-auto-launch 1 %s -o %t && %t
// clang-format on

#include <cudaq.h>

template <std::size_t N>
struct ghz {
  auto operator()() __qpu__ {
    cudaq::qarray<N> q;
    h(q[0]);
    for (int i = 0; i < N - 1; i++) {
      x<cudaq::ctrl>(q[i], q[i + 1]);
    }
    mz(q);
  }
};

int main() {
  {
    auto kernel = ghz<10>{};
    auto counts = cudaq::sample(kernel);
    counts.dump();
    assert(counts.size() == 2);
  }
  {
    // Kernels as lambda functions
    const auto ghz = [](std::size_t N) __qpu__ {
      cudaq::qvector q(N);
      h(q[0]);
      for (int i = 0; i < N - 1; i++) {
        x<cudaq::ctrl>(q[i], q[i + 1]);
      }
      mz(q);
    };
    auto counts = cudaq::sample(ghz, 3);
    counts.dump();
    assert(counts.size() == 2);
  }
  return 0;
}

File: python/tests/ptsbe/sample.py
# ============================================================================ #
# Copyright (c) 2026 NVIDIA Corporation & Affiliates.                          #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #
import pytest
import cudaq
from .kernels import bell


def test_ptsbe_sample_returns_sample_result(depol_noise):
    strategy = cudaq.ptsbe.ExhaustiveSamplingStrategy()
    result = cudaq.ptsbe.sample(bell,
                                noise_model=depol_noise,
                                shots_count=100,
                                sampling_strategy=strategy)
    assert isinstance(result, cudaq.SampleResult)
    assert len(result) > 0


def test_ptsbe_sample_custom_shots(depol_noise):
    strategy = cudaq.ptsbe.ExhaustiveSamplingStrategy()
    result = cudaq.ptsbe.sample(bell,
                                noise_model=depol_noise,
                                shots_count=50,
                                sampling_strategy=strategy)
    assert isinstance(result, cudaq.SampleResult)
    assert len(result) > 0


def test_ptsbe_sample_max_trajectories(depol_noise):
    strategy = cudaq.ptsbe.ExhaustiveSamplingStrategy()
    result = cudaq.ptsbe.sample(
        bell,
        noise_model=depol_noise,
        shots_count=100,
        max_trajectories=50,
        sampling_strategy=strategy,
    )
    assert isinstance(result, cudaq.SampleResult)
    assert len(result) > 0


def test_ptsbe_sample_raises_without_noise_model():
    with pytest.raises(RuntimeError, match="requires a noise_model"):
        cudaq.ptsbe.sample(bell)


def test_ptsbe_sample_raises_with_none_noise_model():
    with pytest.raises(RuntimeError, match="requires a noise_model"):
        cudaq.ptsbe.sample(bell, noise_model=None)


def test_ptsbe_sample_rejects_negative_shots(depol_noise):
    with pytest.raises(RuntimeError, match="shots_count"):
        cudaq.ptsbe.sample(bell, noise_model=depol_noise, shots_count=-1)


def test_ptsbe_sample_rejects_wrong_arity(depol_noise):
    with pytest.raises(RuntimeError, match="Invalid number of arguments"):
        cudaq.ptsbe.sample(bell, 42, noise_model=depol_noise)


def test_ptsbe_sample_rejects_zero_max_trajectories(depol_noise):
    with pytest.raises(RuntimeError, match="max_trajectories"):
        cudaq.ptsbe.sample(bell, noise_model=depol_noise, max_trajectories=0)


def test_ptsbe_sample_rejects_negative_max_trajectories(depol_noise):
    with pytest.raises(RuntimeError, match="max_trajectories"):
        cudaq.ptsbe.sample(bell, noise_model=depol_noise, max_trajectories=-5)
